code/modules/antagonists/traitor/contractor/datums/syndicate_contract.dm

```diff
@@ -0,0 +1,518 @@
+	)))
+	owning_hub.current_contract = null
+
+	status = CONTRACT_STATUS_COMPLETED
+	completed_time = station_time_timestamp()
+	dead_extraction = target_dead
+	addtimer(CALLBACK(src, .proc/notify_completion, final_tc_reward, reward_credits, target_dead), COMPLETION_NOTIFY_DELAY)
+
+/**
+  * Marks the contract as invalid and effectively cancels it for later use.
+  */
+/datum/syndicate_contract/proc/invalidate()
+	if(status in list(CONTRACT_STATUS_COMPLETED, CONTRACT_STATUS_FAILED))
+		return
+
+	clean_up()
+
+	var/pre_text = ""
+	if(src == owning_hub.current_contract)
+		owning_hub.current_contract = null
+		pre_text = "Agent, it appears the target you were tasked to kidnap can no longer be reached."
+	else
+		pre_text = "Agent, a still inactive contract can no longer be done as the target has gone off our sensors."
+
+	var/outcome_text = ""
+	if(generate())
+		status = CONTRACT_STATUS_INACTIVE
+		outcome_text = "Luckily, there is another target on station we can interrogate. A new contract can be found in your uplink."
+	else
+		// Too bad.
+		status = CONTRACT_STATUS_INVALID
+		outcome_text = "Unfortunately, we could not find another target to interrogate and thus we cannot give you another contract."
+
+	if(owning_hub.uplink)
+		owning_hub.uplink.message_holder("[pre_text] [outcome_text]", 'sound/machines/terminal_prompt_deny.ogg')
+		SStgui.update_uis(owning_hub.uplink)
+
+/**
+  * Marks the contract as failed and stops it.
+  *
+  * Arguments:
+  * * difficulty - The visual reason as to why the contract failed.
+  */
+/datum/syndicate_contract/proc/fail(reason = "")
+	if(status != CONTRACT_STATUS_ACTIVE)
+		return
+
+	// Update info
+	owning_hub.current_contract = null
+	status = CONTRACT_STATUS_FAILED
+	fail_reason = reason
+	// Notify
+	clean_up()
+	owning_hub.uplink?.message_holder("You failed to kidnap the target, agent. Do not disappoint us again.", 'sound/machines/terminal_prompt_deny.ogg')
+
+/**
+  * Initiates the extraction process if conditions are met.
+  *
+  * Arguments:
+  * * M - The contractor.
+  */
+/datum/syndicate_contract/proc/start_extraction_process(obj/item/contractor_uplink/U, mob/living/carbon/human/M)
+	if(!U?.Adjacent(M))
+		return "Where in space is your uplink?!"
+	else if(status != CONTRACT_STATUS_ACTIVE)
+		return "This contract is not active."
+	else if(extraction_deadline > world.time)
+		return "Another extraction attempt cannot be made yet."
+
+	var/mob/target = contract.target.current
+	if(!target)
+		invalidate()
+		return "The target is no longer on our sensors. Your contract will be invalidated and replaced with another one."
+	else if(!contract.can_start_extraction_process(M, target))
+		return "You and the target must be standing in the extraction area to start the extraction process."
+
+	M.visible_message("<span class='notice'>[M] starts entering a cryptic series of characters on [U].</span>",\
+					  "<span class='notice'>You start entering an extraction signal to your handlers on [U]..</span>")
```

Comment:
.. one to many

Reply:
One missing actually, since it's an action that takes time